pkg/dockerregistry/server/blobdescriptorservice.go

```diff
@@ -25,6 +38,137 @@ type blobDescriptorService struct {
 	distribution.BlobDescriptorService
 }
 
+// Stat returns a a blob descriptor if the given blob is either linked in repository or is referenced in
+// corresponding image stream. This method is invoked from inside of upstream's linkedBlobStore. It expects
+// a proper repository object to be set on given context by upper openshift middleware wrappers.
 func (bs *blobDescriptorService) Stat(ctx context.Context, dgst digest.Digest) (distribution.Descriptor, error) {
-	return dockerRegistry.BlobStatter().Stat(ctx, dgst)
+	// if there is a repo layer link, return its desc
+	desc, err := bs.BlobDescriptorService.Stat(ctx, dgst)
+	if err == nil {
+		return desc, err
+	}
+
+	// verify the blob is stored locally
+	desc, err = dockerRegistry.BlobStatter().Stat(ctx, dgst)
+	if err != nil {
+		return desc, err
+	}
+
+	// verify that an image holding this blob is actually present in the image stream
+	repo, found := RepositoryFrom(ctx)
+	if !found {
+		context.GetLogger(ctx).Errorf("failed to retrieve repository from context")
+		return distribution.Descriptor{}, distribution.ErrBlobUnknown
+	}
+
+	if imageStreamHasBlob(repo, dgst) {
+		return desc, err
+	}
+
+	return distribution.Descriptor{}, distribution.ErrBlobUnknown
+}
+
+// imageStreamHasBlob returns true if the given blob digest is referenced in image stream corresponding to
+// given repository. If not found locally, image stream's images will be iterated and fetched from newest to
+// oldest until found. Each processed image will update local cache of blobs.
+func imageStreamHasBlob(r *repository, dgst digest.Digest) bool {
+	repositories := r.cachedLayers.RepositoriesForDigest(dgst)
+	match := imageapi.DockerImageReference{Namespace: r.namespace, Name: r.name}.Exact()
+	for _, repo := range repositories {
```

Comment:
The access to the manifest is lost once it's untagged. Ideally, access to its blobs should be denied at the same time. However, even if we ditch the cache, we're still using layer links. These won't be removed until pruned. Client will most probably hit the link itself during stat. So proper pruning is of higher priority to deal with this extended access.

Nevertheless the TTL should be configurable indeed. Otherwise, with few enough images, the entries may never get  out of cache until registry is restarted.

I'll add a TTL.